controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/DeploymentMetricsMaintainer.java

```diff
@@ -25,19 +29,32 @@
     protected void maintain() {
 
         for (Application application : controller().applications().asList()) {
-            try (Lock lock = controller().applications().lock(application.id())) {
-                for (Deployment deployment : application.deployments().values()) {
-
+            for (Deployment deployment : application.deployments().values()) {
+                try {
                     MetricsService.DeploymentMetrics metrics = controller().metricsService()
                             .getDeploymentMetrics(application.id(), deployment.zone());
-
                     DeploymentMetrics appMetrics = new DeploymentMetrics(metrics.queriesPerSecond(), metrics.writesPerSecond(),
-                            metrics.documentCount(), metrics.queryLatencyMillis(), metrics.writeLatencyMillis());
-
-                    Application app = application.with(deployment.withMetrics(appMetrics));
-                    controller().applications().store(app, lock);
+                                                                         metrics.documentCount(), metrics.queryLatencyMillis(), metrics.writeLatencyMillis());
+
+                    // Avoid locking for a long time, due to slow YAMAS.
+                    try (Lock lock = controller().applications().lock(application.id())) {
+                        // Deployment (or application) may have changed (or be gone) now:
+                        controller().applications().get(application.id()).ifPresent(freshApplication -> {
+                            Deployment freshDeployment = freshApplication.deployments().get(deployment.zone());
+                            if (freshDeployment != null)
+                                controller().applications().store(freshApplication.with(freshDeployment.withMetrics(appMetrics)), lock);
+                        });
+                    }
+                }
+                catch (UncheckedIOException e) {
+                    log.warning("Timed out talking to YAMAS; retrying in " + maintenanceInterval() + ":\n" + e);
```

Comment:
Use 
log.log(Level.WARNING, "Timed out talking to YAMAS; retrying in " + maintenanceInterval()", e);